### PR TITLE
Fixed #12410 -- Added LineLocatePoint GIS function.

### PR DESCRIPTION
**Pull request by sir-sigurd**
https://code.djangoproject.com/ticket/12410

### CODE REVIEW

File: django/contrib/gis/db/backends/base/operations.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ class BaseSpatialOperations:
     unsupported_functions = {
         'Area', 'AsGeoJSON', 'AsGML', 'AsKML', 'AsSVG',
         'BoundingCircle', 'Centroid', 'Difference', 'Distance', 'Envelope',
-        'ForceRHR', 'GeoHash', 'Intersection', 'IsValid', 'Length', 'MakeValid',
+        'ForceRHR', 'GeoHash', 'Intersection', 'IsValid', 'Length', 'LineLocatePoint', 'MakeValid',
```
Comment (timgraham) [inline django/contrib/gis/db/backends/base/operations.py]: Please wrap these lines at 79 chars.

File: docs/ref/contrib/gis/functions.txt [hunk 2]
```diff
@@ -346,6 +346,17 @@ resource-intensive) with the ``spheroid`` keyword argument.
     In older versions, a raw value was returned on MySQL when used on
     projected SRS.
 
+``LineLocatePoint``
+===================
+
+.. class:: LineLocatePoint(linestring, point, **extra)
+
```
Comment (timgraham) [inline docs/ref/contrib/gis/functions.txt]: Add `.. versionadded:: 2.0`

File: docs/ref/contrib/gis/functions.txt [hunk 3]
```diff
@@ -346,6 +346,17 @@ resource-intensive) with the ``spheroid`` keyword argument.
...
+*Availability*: `PostGIS <https://postgis.net/docs/ST_LineLocatePoint.html>`__,
+SpatiaLite
+
+Returns a float between 0 and 1 representing the location of the closest point on
+``linestring`` to the given ``point``, as a fraction of total 2d line length.
```
Comment (timgraham) [inline docs/ref/contrib/gis/functions.txt]: I'm not sure if "total 2d line length" needs a grammar tweak or if I'm not familiar with the terminology. Maybe adding "the total" would help? Possibly 2d should be 2D for consistency with other docs?

File: docs/releases/2.0.txt [hunk 4]
```diff
@@ -62,7 +62,8 @@ Minor features
 :mod:`django.contrib.gis`
 ~~~~~~~~~~~~~~~~~~~~~~~~~
 
-* ...
+* Added :class:`~django.contrib.gis.db.models.functions.LineLocatePoint` function
```
Comment (timgraham) [inline docs/releases/2.0.txt]: Added the
function, supported (add comma)

File: tests/gis_tests/geoapp/test_functions.py [hunk 5]
```diff
@@ -279,6 +279,11 @@ def test_area_lookups(self):
         with self.assertRaisesMessage(ValueError, 'AreaField only accepts Area measurement objects.'):
             qs.get(area__lt=500000)
 
+    @skipUnlessDBFeature("has_LineLocatePoint_function")
+    def test_line_locate_point(self):
+        pos_expr = functions.LineLocatePoint(LineString((0, 0), (0, 3), srid=4326), Point(0, 1, srid=4326))
+        self.assertAlmostEquals(State.objects.annotate(pos=pos_expr).first().pos, 0.3333333)
```
Comment (timgraham) [inline tests/gis_tests/geoapp/test_functions.py]: I guess this is like `assertEqual` where the "Equals" version is a deprecated alias.

File: docs/ref/contrib/gis/functions.txt (hunk 3 above)
Comment (sir-sigurd) [inline docs/ref/contrib/gis/functions.txt]: I copied this from https://postgis.net/docs/ST_LineLocatePoint.html
Does "fraction of the 2D line length" look better?

Comment (timgraham) [inline docs/ref/contrib/gis/functions.txt]: Yes

**Pull request merged**